### ISSUE TITLE
Support for RAML 1.0

### ISSUE THREAD

**Issue by lucassimaoo** [mulesoft-labs/raml-for-jax-rs#150]
A new version of RAML spec was released recently (http://docs.raml.org/specs/1.0/). Is there any plan to implement it?

Thanks

**Comment by ddidier** [comment]
+1

**Comment by lufeihaidao** [comment]
+1

**Comment by thenanox** [comment]
So something about this?

**Comment by jantoniucci** [comment]
+1

**Comment by fjavierjimenez** [comment]
+1

**Comment by jeroenver** [comment]
I guess we need to wait for a release of https://github.com/raml-org/raml-java-parser-2 ?

**Comment by cottreda** [comment]
It might be super groovy to put something in the documentation here about the lack of support for RAML 1.0 as well.

**Comment by justjacksonn** [comment]
There is still work on RAML 1.0 going on, so give it some time. RAML2JAXRS with 1.0 support will be worked on once the 1.0 parsers are working correctly.

**Comment by marstein** [comment]
+1 same here.
If we need to wait for https://github.com/raml-org/raml-java-parser-2 then it may be a long wait. Last work on that was 4 months ago. But maybe they develop elsewhere and merge in?

**Comment by r-alukhanov** [comment]
+1

**Comment by ddossot** [comment]
Any progress on this? @sichvoge?  RAML 1.0 is almost 6 months old now, its adoption is conditional to tools supporting it...

**Comment by sichvoge** [comment]
@ddossot RAML 1.0 is currently in its RC stage. Therefore, you see that there is a delay in tooling support since things can change from candidate to candidate.

**Comment by ddossot** [comment]
@sichvoge Roger that, thanks a lot for chiming in and clarifying :+1:

**Comment by sichvoge** [comment]
Sure no problem. I'll ensure that this will be part of the roadmap.

**Comment by schwickster** [comment]
RC2 got released last week and according to the blog there's a feature stop for V1.0.0. Only things that they will do is fix wordings and such.
V2 (alpha) of the java parser is also being worked on: https://github.com/raml-org/raml-java-parser/tree/v2
Hope to see RAML 1.0 (RC2) support for this project soon 👍

**Comment by PeteTh** [comment]
Any update on RAML 1.0 support ?

**Comment by luizmendesalmeida** [comment]
+1

**Comment by justjacksonn** [comment]
I am possibly looking to do some 1.0 work as I need this as well. No
timeline for me though as it is currently a side project.

**Comment by zyxer** [comment]
I can't find any roadmap related to this project, therefore arise question:
What's the current status of RAML 1.0 support? @sichvoge?

**Comment by andjarnic** [comment]
I dont think anyone has started work on this at this time. I briefly started my own version of this, not sure it will match/fit this particular project or not, as I am starting from scratch myself. If I do get something working, I will see if it makes sense to have it stored here as a 1.0, otherwise I may have a different project/name for it

**Comment by vijayrcse** [comment]
Hi All

I am trying the code gen for RAML to JAX-RS using the raml-jaxrs-maven-plugin.
Can someone please advise if RAML 1.0 version is supported?

My RAML spec is newly written which uses 1.0 features like data types, annotations, multiple example etc.

Please advise.

Thanks & Regards
-Vijay

**Comment by kevinduffey** [comment]
Not yet. I started on a RAML 1.0 to JAXRS as I need it as well. But I am limited on time. I want to avoid using JSON and stick to RAML Types for my attempt at it, but it will take some time before I can put something together.

**Comment by petrochenko-pavel-a** [comment]
Also started to play with RAML 1.0 support.  I was able to setup very basic Java->RAML 1.0 emmitting in my fork.(able to generate reasonable output for countries example). But it is still in baby stage. At the same time I have feeling that RAML->Java interoperability has a broader usage and should be separated in a separate project, as well as a feeling that `raml-for-jax-rs` needs some more love/refactoring/cleanup.

Regards,
Pavel

**Comment by kdabir** [comment]
I am starting a new project and RAML 1.0 looks pretty solid but not being able to generate JAX-RS for the spec is a showstopper :(

It looks like the RAML 1.0's parser, which is called parser-2, it's version 1.0 is released  (confusing versions, huh?)
GH Releases: https://github.com/raml-org/raml-java-parser/releases
Maven Central: https://repo1.maven.org/maven2/org/raml/raml-parser-2/1.0.0/

I want to use RAML 1.0 to generate server stubs. As I don't see raml-1.0 -> jax-rs coming in short time, any alternatives that community might suggest? generating to JAX-RS is not a requirement, anything that can run on JVM would do.

**Comment by kdabir** [comment]
Looks like this is something in the direction i was asking for https://github.com/mulesoft-labs/raml-generator

**Comment by sichvoge** [comment]
We are working on a plan that will add the support to this library; there is one missing piece that we need to get done before that though. Its around emitting from Java. We have someone from the community working on that already. Please stay tuned. ;)

**Comment by keski** [comment]
+1

**Comment by drath3000** [comment]
+1

**Comment by idefixcert** [comment]
+1
Is there something we can actually look at?

**Comment by sichvoge** [comment]
There is a PR for RAML 1.0 and the author is looking for feedback on the implementation. So please go ahead and test, and let him know if you run into problems.

**Comment by petrochenko-pavel-a** [comment]
Hi guys, you can checkout alpha version from my fork of raml-for-jaxrs https://github.com/petrochenko-pavel-a/raml-for-jax-rs/tree/raml1.0.

Any feedback , issues, feature suggestions are very appreciated.

Regards,
Pavel

**Comment by vijayrcse** [comment]
Pavel

I am getting the below validation error
`java.lang.IllegalArgumentException: Invalid RAML definition:
        ERROR Unknown key: standardResourceTypes (line 6, col 2 to 23)
        ERROR Unknown key: standardDataTypes (line 7, col 2 to 19)
        ERROR Unknown key: standardTraits (line 8, col 2 to 16)
        ERROR Unknown key: myCustomTypes (line 9, col 2 to 15)
        ERROR Unknown key: uses (line 5, col 0 to 4)
        ERROR resource type not defined: standardResourceTypes.collection
        at org.raml.jaxrs.codegen.core.AbstractGenerator.run(AbstractGenerator.java:794)
`
Basically my raml spec contains uses:

`#%RAML 1.0
title: Remote Vending API
version: v1.0
baseUri: http://remote-vending/api
mediaType: application/json
uses:
  standardResourceTypes: ./resourceTypes/standardResourceTypes.raml
  standardDataTypes: ./dataTypes/standardDatatypes.raml
  standardTraits: ./traits/standardTraits.raml
  myCustomTypes: ./dataTypes/salesTypes.raml`

**Comment by petrochenko-pavel-a** [comment]
This is pretty weird according to what I see, from the issue, it seems that RAML Java parser is not able to parse your RAML file by some reason.  :-(.  I will try to check how it handles similar situations later today.

Hopefully it is not getting broken from relative include or from resource types declared in libraries....
@sichvoge .

Regards,
Pavel

**Comment by petrochenko-pavel-a** [comment]
P.S. thanks a lot for your report.
If you can provide more details about your RAML (ideally small failing example) it will be easier to debug what happens.

Thanks in advance,
Pavel

**Comment by vijayrcse** [comment]
Let me give a simple one. It fails for a raml file just having dataTypes only.


`#%RAML 1.0
title: My API with Types
mediaType: application/json
types:
  Org:
    type: object
    properties:
      onCall: AlertableAdmin
      Head: Manager
  Person:
    type: object
    properties:
      firstname: string
      lastname:  string
      title?:    string
  Phone:
    type: string
    pattern: "[0-9|-]+"
  Manager:
    type: Person
    properties:
      reports: Person[]
      phone:  Phone
  Admin:
    type: Person
    properties:
      clearanceLevel:
        enum: [ low, high ]
  AlertableAdmin:
    type: Admin
    properties:
      phone: Phone
  Alertable: Manager | AlertableAdmin
/orgs/{orgId}:
  get:
    responses:
      200:
        body:
          application/json:
            type: Org`

**Comment by petrochenko-pavel-a** [comment]
looking

**Comment by petrochenko-pavel-a** [comment]
Hm, this one works perfectly for me.

**Comment by vijayrcse** [comment]
Hey Pavel ,
I checked out https://github.com/petrochenko-pavel-a/raml-for-jax-rs/tree/raml1.0. and then did a mvn install on that

After that I tried code gen option through command line as well as maven-plugin in elicpse. Both gave same result

I will retry this again and confirm one more time to validate that I am using the right version

**Comment by vijayrcse** [comment]
It worked. The simple raml file that I gave you yesterday works fine for code gen

I will continue to test our actual raml spec on 1.0 ( which is using libraries refered via relative include )

Thanks a lot for your time to respond my queries.

Also Can u tell me where I can download the below snapshot ? Could not find in any repo and mvn install could not find it.

com.onpositive.aml:org.aml.java2raml:jar:0.0.8-SNAPSHOT

**Comment by petrochenko-pavel-a** [comment]
>It worked. The simple raml file that I gave you yesterday works fine for code gen
Glad to hear this

>I will continue to test our actual raml spec on 1.0 ( which is using libraries refered via relative include )
Thanks a lot, tests and issues/feature suggestions is what I need most.

>Also Can u tell me where I can download the below snapshot ? Could not find in any repo and mvn install could not find it.

Ups, thanks for reporting this, I am publishing releases to mvn central, but I did not configured CI to publish snapshots properly(will fix it asap - just filed https://github.com/OnPositive/aml/issues/13).

PS. For now snapshots can be installed from by cloning `https://github.com/OnPositive/aml` and running `mvn install`

P.P.S.  Great thanks for being early adopter.
Regards,
Pavel

**Comment by vijayrcse** [comment]
Thanks Pavel

Now I tried with my actual spec that uses resourceTypes, traits , libraries etc

Getting the below exception. Do you recognize this or know why this would occur ?( My RAML spec is verified in ATOM no errors reported )

`java.lang.NullPointerException
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.getValue(TopLevelRamlModelBuilder.java:581)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildUsesMaps(TopLevelRamlModelBuilder.java:84)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.lambda$11(TopLevelRamlModelBuilder.java:92)
        at java.util.ArrayList.forEach(ArrayList.java:1249)
        at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1080)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.lambda$0(TopLevelRamlModelBuilder.java:85)
        at java.util.Optional.ifPresent(Optional.java:159)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildUsesMaps(TopLevelRamlModelBuilder.java:85)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.lambda$11(TopLevelRamlModelBuilder.java:92)
        at java.util.ArrayList.forEach(ArrayList.java:1249)
        at java.util.Collections$UnmodifiableCollection.forEach(Collections.java:1080)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.lambda$0(TopLevelRamlModelBuilder.java:85)
        at java.util.Optional.ifPresent(Optional.java:159)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildUsesMaps(TopLevelRamlModelBuilder.java:85)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.build(TopLevelRamlModelBuilder.java:567)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.build(TopLevelRamlModelBuilder.java:594)
        at org.raml.jaxrs.codegen.core.AbstractGenerator.run(AbstractGenerator.java:772)
        at org.raml.jaxrs.codegen.core.GeneratorProxy.run(GeneratorProxy.java:41)
        at org.raml.jaxrs.codegen.core.Launcher.main(Launcher.java:82)
`

**Comment by petrochenko-pavel-a** [comment]
According to what I see from stack trace it got broken during library parsing. (According to the code, I beleive that it may happen in the cases like)
```raml
uses:
  someLibrary: 
  anotherLibrary:
```
Strangely this does not creates errors in API Workbench.

**Comment by petrochenko-pavel-a** [comment]
Switched dependencies to 0.0.8 release. Added one more example of RAML->JAVA. @vijayrcse do you still experiencing an issue with  TopLevelRamlModelBuilder?

**Comment by vijayrcse** [comment]
Pavel, i will re-test and confirm this today. could not reply as i was out for past 3 days

**Comment by petrochenko-pavel-a** [comment]
Hi guys, I have published raml->jaxrs and raml->jaxrs maven plugin under `com.onpositive.aml` group in maven central for now. So if you would like to test it you may change `<groupId>` value to `com.onpositive.aml` and version to 1.3.6 (just like in the sample configuration below).

### RAML to JAXRS
```xml
<plugin>
	<groupId>com.onpositive.aml</groupId> 
	<artifactId>raml-jaxrs-maven-plugin</artifactId>
	<version>1.3.6</version>
	<configuration>
					<sourceDirectory>${basedir}/src/main/resources/raml</sourceDirectory>
					<basePackageName>org.raml.jaxrs.example</basePackageName>
					<jaxrsVersion>2.0</jaxrsVersion>
					<useJsr303Annotations>false</useJsr303Annotations>
					<jsonMapper>jackson2</jsonMapper>
					<removeOldOutput>true</removeOldOutput>
	</configuration>
	<executions>
		<execution>
			<goals>
	  		<goal>generate</goal>
	         	</goals>
	<phase>generate-sources</phase>
	</execution>
	</executions>
</plugin>
```
### JAXRS->RAML
```xml
			<plugin>
				<version>1.3.6</version>
				<groupId>com.onpositive.aml</groupId>
				<artifactId>jaxrs-raml-maven-plugin</artifactId>
				<configuration>
					<sourceDirectory>${basedir}/src/main/java</sourceDirectory>
					<outputFile>${basedir}/target/generated-sources/jaxrs-raml/tests1/api.raml</outputFile>
					<basePackageName>com.mulesoft.jaxrs.raml.annotation.tests</basePackageName>
					<jaxrsVersion>2.0</jaxrsVersion>
					<useJsr303Annotations>false</useJsr303Annotations>
					<jsonMapper>jackson2</jsonMapper>
					<removeOldOutput>true</removeOldOutput>
				</configuration>
				<executions>
					<execution>
						<goals>
							<goal>generate-raml</goal>
						</goals>
						<phase>test</phase>
					</execution>
				</executions>
			</plugin>
```
Regards,
Pavel

**Comment by vijayrcse** [comment]
I removed all the relative URL and re-tried my RAML spec( with above release plugin )

`
Caused by: java.lang.NullPointerException
        at org.aml.typesystem.AbstractType.addMeta(AbstractType.java:214)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.innerBuild(TopLevelRamlModelBuilder.java:336)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildType(TopLevelRamlModelBuilder.java:240)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildType(TopLevelRamlModelBuilder.java:236)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.resolveType(TopLevelRamlModelBuilder.java:230)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildSuperType(TopLevelRamlModelBuilder.java:208)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.innerBuild(TopLevelRamlModelBuilder.java:292)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildType(TopLevelRamlModelBuilder.java:240)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.parseProperties(TopLevelRamlModelBuilder.java:432)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.innerBuild(TopLevelRamlModelBuilder.java:314)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildType(TopLevelRamlModelBuilder.java:240)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildType(TopLevelRamlModelBuilder.java:236)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.resolveType(TopLevelRamlModelBuilder.java:230)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.lambda$buildTypes$9(TopLevelRamlModelBuilder.java:126)
        at java.util.HashMap$KeySet.forEach(HashMap.java:932)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildTypes(TopLevelRamlModelBuilder.java:126)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.lambda$buildTypes$8(TopLevelRamlModelBuilder.java:125)
        at java.util.LinkedHashMap$LinkedValues.forEach(LinkedHashMap.java:608)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildTypes(TopLevelRamlModelBuilder.java:125)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.build(TopLevelRamlModelBuilder.java:564)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.build(TopLevelRamlModelBuilder.java:590)
        at org.raml.jaxrs.codegen.core.AbstractGenerator.run(AbstractGenerator.java:772)
        at org.raml.jaxrs.codegen.core.GeneratorProxy.run(GeneratorProxy.java:41)
        at org.raml.jaxrs.codegen.maven.RamlJaxrsCodegenMojo.execute(RamlJaxrsCodegenMojo.java:297)
        ... 22 more
`
If I have to give a sample I have to trim down a bit and share. Will let you know

**Comment by vijayrcse** [comment]
Found where it is failing. It fails for below dataTypes declaration

`types:
  Exchange:
    type: object
    properties:
      value: integer
      in: integer
      out: integer
  Sale:
    type: object
    properties:
      machineId: string
      trayId: string
      dateAndTime: datetime-only
      exchange: Exchange
  SaleDetail:
    type: object
    properties:
      dateAndTime: datetime-only
      value: integer
      machineId: string
      productId: string
  SaleDetails:
    type: array
    items: SaleDetail
  SalesSummary:
    type: object
    properties:
      count: integer
      totalValue: integer
      sales: SaleDetails
`

But think this is correct ATOM doesn't throw error. it validates the sample json correctly. any thoughts ?

**Comment by vijayrcse** [comment]
I think below is the one breaking

`
SaleDetail:
    type: object
    properties:
      dateAndTime: datetime-only
      value: integer
      machineId: string
      productId: string
  SaleDetails:
    type: array
    items: SaleDetail
`

But I think this is valid per RAML spec

**Comment by petrochenko-pavel-a** [comment]
Yep, you are totally right, problem is in items+array combination. Fixing it.

**Comment by petrochenko-pavel-a** [comment]
Published 1.3.7 and 0.0.9 version of AML (this should work fine with them)

**Comment by vijayrcse** [comment]
Okay thanks a lot will test again
Also i am going to test today with relative path for "uses" with some "traits"includes
Will let you know

**Comment by vijayrcse** [comment]
I tested the 1.3.7 version. I don't get any errors now. But the code gen output generates one file under resource folder

`
import javax.ws.rs.Path;

@Path("sales")
public interface SalesThatAreDoneSoFarResource {


}
`

It doesn't have any method.

Will share the sample RAML spec.

**Comment by petrochenko-pavel-a** [comment]
Ho Ho. Looking forward to see it

19 Окт 2016 г. 6:11 PM пользователь "vijayrcse" <notifications@github.com>
написал:

**Comment by vijayrcse** [comment]
Try to simplify a bit, and got an error this time

`
#%RAML 1.0
title: My API with Types
mediaType: application/json
uses:
  standardDataTypes: standardDataTypes.raml
  simpleResourceTypes: simpleResourceTypes.raml
/orgs/{orgId}:
  type:
    simpleResourceTypes.collection:
      itemType: standardDataTypes.Org
`

**standardDataTypes.raml**

`
#%RAML 1.0 Library
usage: Standard Data types

types:
Person:
    type: object
    properties:
      firstname: string
      lastname:  string
      title?:    string
  Phone:
    type: string
    pattern: "[0-9|-]+"
  Manager:
    type: Person
    properties:
      reports: Person[]
      phone:  Phone
  Admin:
    type: Person
    properties:
      clearanceLevel:
        enum: [ low, high ]
  AlertableAdmin:
    type: Admin
    properties:
      phone: Phone
  Alertable: Manager | AlertableAdmin
  Org:
    type: object
    properties:
      onCall: AlertableAdmin
      Head: Manager
`
**simpleResourceTypes.raml**
`
#%RAML 1.0 Library
usage: Standard resourceTypes
uses:
  standardDataTypes: standardDataTypes.raml
resourceTypes:
  collection:
    get:
      responses:
        200:
          body:
            application/json:
              type: <<itemType>>
`

**Comment by petrochenko-pavel-a** [comment]
Ok, I know reason, Java parser was not able to resolve include. But still interesting. Working on it.

**Comment by vijayrcse** [comment]
One more question :  is raml-for-jax-rs project supports java client code generation for a given RAML spec ?

We see it only generates an interface

**Comment by petrochenko-pavel-a** [comment]
Theoretically yes. (this means that it actually contains code required to generate inerfaces for jersey clients), in practice it is something like a no, because this code is not exposed to maven plugins, not tested well, e.t.c.

On the good side it does not looks a big job to build a client gen, after foundation is stable, and it is actualy inside of my research plan.

Regards,
Pavel

**Comment by vijayrcse** [comment]
HI Pavel

Any findings on the java parser error ?

**Comment by petrochenko-pavel-a** [comment]
Yep. I fixed all issues that I was able to reproduce. (however I still does not know reason of your original issue (empty generated interface).

I will release next version in ~20 mins.

**Comment by petrochenko-pavel-a** [comment]
QQ: are you ok with installing from git? If yes I would prefer to avoid doing to lot of releases.

**Comment by vijayrcse** [comment]
okay I can try that as well. Let me know the git URL.

Also for the original issue( empty generated interface ) I will try to reproduce after testing the fix. That issue was produced when I test against my RAML spec ( which is quite big compared to wat I shared yesterday ). Just taking a step-by-step approach so it is easier to find and fix problem

**Comment by petrochenko-pavel-a** [comment]
Just published 1.3.8 to maven (but it still not visible in repo)

```
clone https://github.com/petrochenko-pavel-a/raml-for-jax-rs.git
switch to raml1.0
mvn install
```

you will also need aml 0.0.10 (it is already published but may be not accessible yet)

It is located at (https://github.com/OnPositive/aml)

Regards,
Pavel.

**Comment by vijayrcse** [comment]
thanks checking this out

**Comment by vijayrcse** [comment]
getting this now for the same raml i shared above. Not using any relative URL to include libraries.

`
Caused by: java.lang.NullPointerException
        at org.aml.typesystem.InheritedType.addSuperType(InheritedType.java:42)
        at org.aml.typesystem.InheritedType.<init>(InheritedType.java:31)
        at org.aml.typesystem.TypeOps.derive(TypeOps.java:31)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.innerBuild(TopLevelRamlModelBuilder.java:301)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildType(TopLevelRamlModelBuilder.java:246)
        at org.aml.typesystem.ramlreader.AbstractWrappedNodeImpl.toParamList(AbstractWrappedNodeImpl.java:90)
        at org.aml.typesystem.ramlreader.AbstractWrappedNodeImpl.toMimeTypes(AbstractWrappedNodeImpl.java:38)
        at org.aml.typesystem.ramlreader.AbstractWrappedNodeImpl.body(AbstractWrappedNodeImpl.java:48)
        at org.aml.typesystem.ramlreader.AbstractWrappedNodeImpl.hasBody(AbstractWrappedNodeImpl.java:58)
`
were you able to test against my raml that i shared before ?

**Comment by petrochenko-pavel-a** [comment]
yes, I even added it to tests.

**Comment by petrochenko-pavel-a** [comment]
Just pushed more accurate issue reporting to aml.  (does not releasing anything), so you will need to update aml from git.

**Comment by vijayrcse** [comment]
okay let me try

**Comment by vijayrcse** [comment]
can you tell me know what this error means ?

Caused by: java.lang.IllegalArgumentException: At least one ancestor is needed
        at org.aml.typesystem.TypeOps.derive(TypeOps.java:29)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.innerBuild(TopLevelRamlModelBuilder.java:303)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildType(TopLevelRamlModelBuilder.java:246)
        at org.aml.typesystem.ramlreader.AbstractWrappedNodeImpl.toParamList(AbstractWrappedNodeImpl.java:90)
        at org.aml.typesystem.ramlreader.AbstractWrappedNodeImpl.toMimeTypes(AbstractWrappedNodeImpl.java:38)

**Comment by petrochenko-pavel-a** [comment]
This means that typesystem attempts to extend a type when no super type is specified, it is definitely my fault. Most probably error in default type calculation, give me a sec.

**Comment by petrochenko-pavel-a** [comment]
Do you have a sample to reproduce?

**Comment by vijayrcse** [comment]
will give in a min. but it is same raml

**Comment by petrochenko-pavel-a** [comment]
buh, it works fine for me.

**Comment by vijayrcse** [comment]
could be due to myself pointing to wrong version of any libs. i will validate from my side

api : simple.raml

these are the files
[simple.raml.txt](https://github.com/mulesoft/raml-for-jax-rs/files/541676/simple.raml.txt)
[simpleResourceTypes.raml.txt](https://github.com/mulesoft/raml-for-jax-rs/files/541677/simpleResourceTypes.raml.txt)
[standardDataTypes.raml.txt](https://github.com/mulesoft/raml-for-jax-rs/files/541678/standardDataTypes.raml.txt)

**Comment by vijayrcse** [comment]
I am getting similar error from git code as well

testcase : RAML1Test.java

`
Tests run: 2, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 1.146 sec <<< FAILURE!
runComplete(org.raml.jaxrs.codegen.core.RAML1Test)  Time elapsed: 0.822 sec  <<< ERROR!
java.lang.IllegalArgumentException: At least one ancestor is needed
        at org.aml.typesystem.TypeOps.derive(TypeOps.java:29)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.innerBuild(TopLevelRamlModelBuilder.java:303)
        at org.aml.typesystem.ramlreader.TopLevelRamlModelBuilder.buildType(TopLevelRamlModelBuilder.java:246)

`

**Comment by petrochenko-pavel-a** [comment]
yep, this was my fault. Fixing. this one + another one.

P.S. thanks a lot for your help and a big sorry for this issues (working on another project simultaniously)

**Comment by petrochenko-pavel-a** [comment]
Fixed, works fine for me now. (please pull and test on your side to)

**Comment by vijayrcse** [comment]
doing right away

**Comment by vijayrcse** [comment]
excellent!. it worked :+1:

**Comment by vijayrcse** [comment]
I will continue to test adding back additional features from my original RAML 1.,0 spec. will let you know if i come across anything..

**Comment by petrochenko-pavel-a** [comment]
Thanks a lot.!!!
Hopefully we are near to getting it working with your spec.

Regards,
Pavel

**Comment by dmiorandi** [comment]
Hi everybody, I'm interesting ot migrate my actual raml 0.8 + jsonschema draft3 projects (so jsonschema2pojo as base)
Just some questions to understand:

Does new raml1.0/new java code generator supports:
* java extends 
* java implements
* Json schema draft3
* custom field type binding (i.e. map numeric field to BigDecimal, java 8 date datetime localdate, ...)
* custom object / class type mapping (on a specific java class / package)

Is it also possible to convert raml types/objects to jsonschema draft 3 or 4 output with object id's. (A tools I use to generate d.ts from jsonschema require this)?

Any other big features on jsonschema2pojo not supported here?
For sure writing jsonschema is difficult but jsonschema2pojo is a mature project so I would to understand actual situation. Tks everybody.

To be clear this is an example:
so in your opinion are there something not working on raml 1.0 + java converter, using raml types?
```
ApiEurisingMgrAgenPilotProjects: |
    {  "$schema": "http://json-schema.org/draft-03/schema",
      "id": "ApiEurisingMgrAgenPilotProjects",
      "type": "object",
      "extends" : {"$ref" : "ApiBaseObject"},
      "javaType" : "it.clesius.eurising.eurising_mgr.api.model.ApiEurisingMgrAgenPilotProjects",
      "description": "",
      "properties": {
          "idPilotProject" : { "type": "integer", "required": true },
          "projectModel" : { "type": "string", "required": true },
          "resolutionType" : { "$ref": "ApiEurisingMgrAgenResolutionTypes", "required": true },
          "resolutionNr" : { "type": "string" },
          "resolutionDate" : { "type" : "string", "format" : "date" },
          "amount" :  { "type" : "number", "multipleOf": 0.01, "javaType": "java.math.BigDecimal" }
      }
    }
```

**Comment by vijayrcse** [comment]
Pavel 
Do you think resource type inheritance is supported ?

`
resourceTypes:
  single:
    get:
      responses:
        200:
          body:
            application/json:
              type: <<itemType>>

  read-only-collection:
    type: { single: { itemType : <<itemType>> } }
    get:
      responses:
        200:
          body:
            application/json:
              type: <<itemType>>[]
              example: <<exampleCollection>>
`

It breaks at type:...

[ERROR] Invalid element {
[ERROR] type: <<itemType>>
[ERROR] }.
[ERROR] Invalid element {
[ERROR] type: <<itemType>>
[ERROR] }.

**Comment by petrochenko-pavel-a** [comment]
No, I have just tried basic sample with resource type inheritance and it works fine. According to what I see from RAML Parser output it looks like it failed to substitute  type parameter. So I need more info (sample of resource type appication to debug)

Regards,
Pavel

**Comment by petrochenko-pavel-a** [comment]
Hi @dmiorandi 

In a few words answer is partially you may look in the list of supported features at [AML RAML->Java](https://github.com/OnPositive/aml/blob/master/raml2java.md) some of them are already wired to raml-for-jaxrs configuration, some are not yet.

If you need something particular please file an issue/feature request and I will try to handle it asap.

>Is it also possible to convert raml types/objects to jsonschema draft 3 or 4 output with object id's. (A tools I use to generate d.ts from jsonschema require this)?

There is something in JS world (https://github.com/raml-org/ramldt2jsonschema) but I guess nothing exists on Java Side.

P.S. (will review in a more details later, according to what I see now, not everything will work from scratch but it should not require a lot of time to support missed pieces)

Regards,
Pavel

**Comment by dmiorandi** [comment]
Ok, so issues was created. This is what I understand from documentation

SUPPORTED
* I can partially or completly use external jsonschema like in raml 0.8 (based on jsonschema2pojo)
* BigDecimal, BigInteger native support, but no Java8 time

MISSING
* Custom naming of java class (class name and package name) instead same name as raml object
  https://github.com/OnPositive/aml/issues/19
* Possibility to set different package for different raml file. I.e. common raml in common package
  business raml in business package https://github.com/OnPositive/aml/issues/18
* Generic java implements / extends support on raml Types (not only serializable & c) https://github.com/OnPositive/aml/issues/17
* Java8 time java.time.* / JSR310. See https://github.com/OnPositive/aml/issues/16

**Comment by petrochenko-pavel-a** [comment]
Thanks a lot @dmiorandi .

Will handle it in nearest days.

Regards,
Pavel
